src/services/jsonLinks.ts

```diff
@@ -6,24 +6,75 @@
 import { DocumentLink } from 'vscode-languageserver-types';
 import { TextDocument, ASTNode, PropertyASTNode, Range, Thenable } from '../jsonLanguageTypes';
 import { JSONDocument } from '../parser/jsonParser';
+import { IJSONSchemaService } from './jsonSchemaService';
+import { URI } from 'vscode-uri';
+import { existsSync as fileExistsSync } from 'fs';
```

Comment:
We want to run that code also in the web, so we can not use NodeJS-API like `fs`.
Instead we need to provide file system access through an abstraction.
The already have that in the css-languageservice: https://github.com/microsoft/vscode-css-languageservice/blob/885c65cd38033a456225dd607a8d07776cd6bb6f/src/cssLanguageTypes.ts#L173

Reply:
Sure! If I'm understanding things correctly, it looks like I'll need to use `stat` (I didn't see an exists function in the [fs provider docs](https://code.visualstudio.com/api/references/vscode-api#FileSystemProvider)) and handle the promise it returns. 

Then when I eventually want to use this in the yaml language server, I'm assuming I'll need to do something similar to [what you did here](https://github.com/redhat-developer/yaml-language-server/pull/539).

Please let me know if I'm not on right path.

Reply:
Looks like my link was corrupted, I updated it. 

What's to do:
- add new type to jsonLanguageTypes.ts:
```
export interface FileSystemProvider {
	stat(uri: DocumentUri): Promise<FileStat>;
}
```
plus `FileStat` and `FileType` (copy these from the [cssLanguageTypes](https://github.com/microsoft/vscode-css-languageservice/blob/3e2d7992179c81dc327c5e1ec8b2ec59cb2220f1/src/cssLanguageTypes.ts#L245C13-L245C21))

- add a new property 
```
	/**
	 * Abstract file system access away from the service.
	 * Used for dynamic link resolving, path completion, etc.
	 */
	fileSystemProvider?: FileSystemProvider;
``` 
to `LanguageServiceParams`
- pass that to your code and use it, if available (all is optional, we don't want to break existing clients)
- document these types in the CHANGELOG.md
- in followup PRs we have to update the consumers of the `vscode-json-languageservice` (yaml-extension and json-language-feature) to provide a FileSystemProvider implementation. I can do that for `vscode-json-languageservice`

Reply:
@aeschli Thanks for the detailed breakdown - made things much easier! I rebased against main and made the suggested changes.

I don't do much JavaScript development, and I'm not 100% happy w/ how complicated all the promise logic ended up where I'm checking if the file exists. If you have any suggestions for simplifying things, I'm all ears 😄.